File: ui/src/reliability_stats.rs
//! Pure reliability-metrics computation for the RELIABILITY page: per-routine success rate,
//! active pass/fail streak, and flakiness (status-flip rate) over each routine's most recent
//! finished runs. Split from `reliability.rs` (the page shell) so the math stays host-testable
//! without a wasm/DOM dependency, mirroring `overview.rs`'s split from `overview_tests.rs`.
//!
//! Best practice (CI/CD reliability dashboards — GitHub Actions Insights, `CircleCI` Insights,
//! Datadog Test Visibility): rank jobs by recent success rate and flag flaky ones (alternating
//! pass/fail) separately from steadily-failing ones, since the two call for different responses
//! — a flaky routine needs investigation, a steadily-failing one needs immediate attention.

use std::cmp::Ordering;
use std::collections::HashMap;

use crate::routines::{FleetRunSummary, RunStatus};

/// Most recent finished runs per routine considered for reliability metrics. Caps memory/render
/// cost per routine the same way `sparkline.rs`'s `SPARKLINE_LEN` caps its tick strip, just with
/// a longer window since flakiness needs more samples than an at-a-glance sparkline does.
pub(crate) const SAMPLE_LEN: usize = 20;

/// Minimum finished-run sample before flakiness is judged — a 2-run sample that flips once is
/// noise, not a signal.
const FLAKY_MIN_SAMPLE: usize = 5;

/// A routine whose adjacent-run status flips make up at least this fraction of its sample's
/// adjacent pairs is flagged flaky.
const FLAKY_FLIP_RATIO: f64 = 0.4;

/// A routine's most-recent-run streak: consecutive finished runs, newest first, sharing the same
/// outcome.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum Streak {
    /// `n` consecutive successes (n >= 1), most recent run first.
    Success(usize),
    /// `n` consecutive failures (n >= 1), most recent run first.
    Failure(usize),
    /// No finished run in the sample.
    None,
}

/// Reliability metrics for one routine, derived from its most recent finished
/// (`Success`/`Failed`) runs — `Running`/`Unknown` runs are excluded since they say nothing
/// about reliability.
#[derive(Debug, Clone, PartialEq)]
pub(crate) struct RoutineReliability {
    pub routine_id: String,
    pub routine_title: String,
    /// Finished runs considered, out of up to `SAMPLE_LEN`.
    pub sample_size: usize,
    /// `Success` runs within the sample.
    pub successes: usize,
    pub streak: Streak,
    /// Count of status flips (success→failure or vice versa) across adjacent runs in the
    /// sample — high relative to `sample_size` indicates flakiness rather than a steady trend.
    pub flips: usize,
}

impl RoutineReliability {
    /// `successes / sample_size`, or `None` when the sample is empty.
    pub(crate) fn success_rate(&self) -> Option<f64> {
        if self.sample_size == 0 {
            None
        } else {
            Some(self.successes as f64 / self.sample_size as f64)
        }
    }

    /// `true` when this routine's flip rate crosses the flaky threshold.
    pub(crate) fn is_flaky(&self) -> bool {
        if self.sample_size < FLAKY_MIN_SAMPLE {
            return false;
        }
        let pairs = self.sample_size - 1;
        pairs > 0 && (self.flips as f64) / (pairs as f64) >= FLAKY_FLIP_RATIO
    }
}

/// Fleet-wide reliability summary, aggregated across every routine's sample.
#[derive(Debug, Clone, PartialEq, Default)]
pub(crate) struct FleetReliability {
    pub sample_size: usize,
    pub successes: usize,
    /// Routines with an active (>= 1 run) failure streak.
    pub failing_count: usize,
    pub flaky_count: usize,
}

impl FleetReliability {
    /// `successes / sample_size`, or `None` when nothing has been sampled yet.
    pub(crate) fn success_rate(&self) -> Option<f64> {
        if self.sample_size == 0 {
            None
        } else {
            Some(self.successes as f64 / self.sample_size as f64)
        }
    }
}

/// Per-routine accumulator while bucketing the fleet-wide run list.
struct RoutineAcc {
    title: String,
    /// Newest-first, capped at `SAMPLE_LEN`.
    statuses: Vec<RunStatus>,
}

/// Buckets a fleet-wide, newest-first run list by routine, keeping only `Success`/`Failed` runs
/// and capping each bucket at `SAMPLE_LEN`, newest-first.
fn bucket_finished_runs(runs: &[FleetRunSummary]) -> HashMap<String, RoutineAcc> {
    let mut by_routine: HashMap<String, RoutineAcc> = HashMap::new();
    for run in runs {
        if !matches!(run.status, RunStatus::Success | RunStatus::Failed) {
            continue;
        }
        let acc = by_routine
            .entry(run.routine_id.clone())
            .or_insert_with(|| RoutineAcc {
                title: run.routine_title.clone(),
                statuses: Vec::new(),
            });
        if acc.statuses.len() < SAMPLE_LEN {
            acc.statuses.push(run.status);
        }
    }
    by_routine
}

/// The active streak at the head (newest-first) of a finished-run sample.
fn compute_streak(statuses: &[RunStatus]) -> Streak {
    let Some(&newest) = statuses.first() else {
        return Streak::None;
    };
    let n = statuses.iter().take_while(|&&s| s == newest).count();
    if newest == RunStatus::Success {
        Streak::Success(n)
    } else {
        Streak::Failure(n)
    }
}

/// Count of adjacent-pair status changes in a finished-run sample.
fn count_flips(statuses: &[RunStatus]) -> usize {
    statuses
        .windows(2)
        .filter(|pair| pair[0] != pair[1])
        .count()
}

/// The length of an active failure streak, or 0 for a success streak or no sample.
fn failure_streak_len(streak: Streak) -> usize {
    match streak {
        Streak::Failure(n) => n,
        Streak::Success(_) | Streak::None => 0,
    }
}

/// Computes every routine's reliability metrics from a fleet-wide run list, ranked
/// worst-first: an active failure streak outranks everything else (longer streak first),
/// then lowest success rate, then title for a stable tie-break. Routines with no finished
/// run in the sample are omitted — there is nothing to rank.
pub(crate) fn compute_reliability(runs: &[FleetRunSummary]) -> Vec<RoutineReliability> {
    let mut items: Vec<RoutineReliability> = bucket_finished_runs(runs)
        .into_iter()
        .map(|(routine_id, acc)| RoutineReliability {
            sample_size: acc.statuses.len(),
            successes: acc
                .statuses
                .iter()
                .filter(|&&s| s == RunStatus::Success)
                .count(),
            streak: compute_streak(&acc.statuses),
            flips: count_flips(&acc.statuses),
            routine_title: acc.title,
            routine_id,
        })
        .collect();

    items.sort_by(|a, b| {
        failure_streak_len(b.streak)
            .cmp(&failure_streak_len(a.streak))
            .then_with(|| {
                let a_rate = a.success_rate().unwrap_or(1.0);
                let b_rate = b.success_rate().unwrap_or(1.0);
                a_rate.partial_cmp(&b_rate).unwrap_or(Ordering::Equal)
            })
            .then_with(|| a.routine_title.cmp(&b.routine_title))
    });
    items
}

/// Aggregates per-routine metrics into a fleet-wide summary for the page's stat tiles.
pub(crate) fn fleet_summary(items: &[RoutineReliability]) -> FleetReliability {
    FleetReliability {
        sample_size: items.iter().map(|r| r.sample_size).sum(),
        successes: items.iter().map(|r| r.successes).sum(),
        failing_count: items
            .iter()
            .filter(|r| failure_streak_len(r.streak) > 0)
            .count(),
        flaky_count: items.iter().filter(|r| r.is_flaky()).count(),
    }
}

/// CSS class for a routine's streak badge (reuses the generic `run-status` pill classes).
pub(crate) fn streak_class(streak: Streak) -> &'static str {
    match streak {
        Streak::Success(_) => "run-status success",
        Streak::Failure(_) => "run-status failed",
        Streak::None => "run-status unknown",
    }
}

/// Display label for a routine's streak badge.
pub(crate) fn streak_label(streak: Streak) -> String {
    match streak {
        Streak::Success(n) => format!("{n} OK"),
        Streak::Failure(n) => format!("{n} FAILING"),
        Streak::None => "—".to_string(),
    }
}

/// CSS class for a success-rate badge (reuses the generic `run-status` pill classes).
pub(crate) fn rate_class(rate: Option<f64>) -> &'static str {
    match rate {
        None => "run-status unknown",
        Some(r) if r >= 0.9 => "run-status success",
        Some(r) if r >= 0.7 => "run-status running",
        Some(_) => "run-status failed",
    }
}

/// Display label for a success-rate badge.
pub(crate) fn rate_label(rate: Option<f64>) -> String {
    match rate {
        None => "—".to_string(),
        Some(r) => format!("{:.0}%", r * 100.0),
    }
}

#[cfg(test)]
#[path = "reliability_stats_tests.rs"]
mod reliability_stats_tests;

File: client/src/pages/reliability/reliabilityStats.ts
/**
 * Pure reliability-metrics computation for the RELIABILITY page: per-routine success rate,
 * active pass/fail streak, flakiness (status-flip rate), and run-duration percentiles with a
 * slower-trend regression flag, over each routine's most recent finished runs.
 *
 * Best practice (CI/CD reliability dashboards — GitHub Actions Insights, CircleCI Insights,
 * Datadog CI Visibility): rank jobs by recent success rate, flag flaky ones (alternating
 * pass/fail) separately from steadily-failing ones, and track p50/p95 duration plus a
 * recent-vs-baseline comparison — p50 shows the typical run, p95 exposes worst-case outliers an
 * average hides, and the trend comparison catches a regression before it escalates into
 * failures. Ports `ui/src/reliability_stats.rs`'s ranking/flakiness math (pre-dating the removal
 * of the Yew UI in favor of this React client) and extends it with the duration dimension.
 */
import type { FleetRunSummary } from "../../api/hooks";

/** Most recent finished runs per routine considered for reliability metrics. */
export const SAMPLE_LEN = 20;

/** Minimum finished-run sample before flakiness is judged. */
const FLAKY_MIN_SAMPLE = 5;

/** Adjacent-pair flip ratio at/above which a routine is flagged flaky. */
const FLAKY_FLIP_RATIO = 0.4;

/** Minimum timed-run sample before a duration regression is judged (needs two stable halves). */
const REGRESSION_MIN_SAMPLE = 8;

/** Newer-half median duration must exceed the older half's by this factor to flag a regression. */
const REGRESSION_RATIO = 1.25;

export type Streak =
  | { kind: "success"; count: number }
  | { kind: "failure"; count: number }
  | { kind: "none" };

export interface RoutineReliability {
  routineId: string;
  routineTitle: string;
  /** Finished runs considered, out of up to `SAMPLE_LEN`. */
  sampleSize: number;
  /** `success` runs within the sample. */
  successes: number;
  streak: Streak;
  /** Count of status flips across adjacent runs in the sample. */
  flips: number;
  /** Finished runs' durations in seconds, newest-first (only runs with both timestamps). */
  durationsSecs: number[];
  p50Secs: number | null;
  p95Secs: number | null;
  /** `true` when this routine's recent runs are meaningfully slower than its baseline. */
  regressing: boolean;
}

export interface FleetReliability {
  sampleSize: number;
  successes: number;
  /** Routines with an active (>= 1 run) failure streak. */
  failingCount: number;
  flakyCount: number;
  p50Secs: number | null;
  p95Secs: number | null;
  regressingCount: number;
}

/** `successes / sampleSize`, or `null` when the sample is empty. */
export function successRate(item: RoutineReliability): number | null {
  return item.sampleSize === 0 ? null : item.successes / item.sampleSize;
}

/** `true` when a routine's flip rate crosses the flaky threshold. */
export function isFlaky(item: RoutineReliability): boolean {
  if (item.sampleSize < FLAKY_MIN_SAMPLE) return false;
  const pairs = item.sampleSize - 1;
  return pairs > 0 && item.flips / pairs >= FLAKY_FLIP_RATIO;
}

/** The length of an active failure streak, or 0 for a success streak or no sample. */
function failureStreakLen(streak: Streak): number {
  return streak.kind === "failure" ? streak.count : 0;
}

/** The active streak at the head (newest-first) of a finished-run status list. */
function computeStreak(statuses: FleetRunSummary["status"][]): Streak {
  const newest = statuses[0];
  if (newest === undefined) return { kind: "none" };
  let n = 0;
  for (const s of statuses) {
    if (s !== newest) break;
    n += 1;
  }
  return newest === "success" ? { kind: "success", count: n } : { kind: "failure", count: n };
}

/** Count of adjacent-pair status changes in a finished-run status list. */
function countFlips(statuses: FleetRunSummary["status"][]): number {
  let flips = 0;
  for (let i = 1; i < statuses.length; i++) {
    if (statuses[i] !== statuses[i - 1]) flips += 1;
  }
  return flips;
}

/** Sample percentile via nearest-rank on an ascending-sorted array (empty input yields `null`). */
function percentile(sortedAsc: number[], p: number): number | null {
  if (sortedAsc.length === 0) return null;
  const idx = Math.min(sortedAsc.length - 1, Math.max(0, Math.ceil(p * sortedAsc.length) - 1));
  return sortedAsc[idx] ?? null;
}

/** Median of a value list, or `null` when empty. */
function median(values: number[]): number | null {
  return percentile([...values].sort((a, b) => a - b), 0.5);
}

/**
 * `true` when a newest-first duration sample's newer half runs meaningfully slower than its
 * older half — the same recent-vs-baseline comparison CI duration-drift tools use to catch a
 * performance regression before it escalates into failures.
 */
function isRegressing(durationsSecs: number[]): boolean {
  if (durationsSecs.length < REGRESSION_MIN_SAMPLE) return false;
  const half = Math.floor(durationsSecs.length / 2);
  const newer = median(durationsSecs.slice(0, half));
  const older = median(durationsSecs.slice(half));
  if (newer === null || older === null || older <= 0) return false;
  return newer > older * REGRESSION_RATIO;
}

interface RoutineAcc {
  title: string;
  /** Newest-first, capped at `SAMPLE_LEN`. */
  statuses: FleetRunSummary["status"][];
  /** Newest-first, capped at `SAMPLE_LEN`; only runs with both timestamps. */
  durationsSecs: number[];
}

/**
 * Buckets a fleet-wide, newest-first run list by routine, keeping only `success`/`failed` runs
 * and capping each bucket's statuses/durations at `SAMPLE_LEN`, newest-first.
 */
function bucketFinishedRuns(runs: FleetRunSummary[]): Map<string, RoutineAcc> {
  const byRoutine = new Map<string, RoutineAcc>();
  for (const run of runs) {
    if (run.status !== "success" && run.status !== "failed") continue;
    let acc = byRoutine.get(run.routine_id);
    if (!acc) {
      acc = { title: run.routine_title, statuses: [], durationsSecs: [] };
      byRoutine.set(run.routine_id, acc);
    }
    if (acc.statuses.length < SAMPLE_LEN) acc.statuses.push(run.status);
    if (acc.durationsSecs.length < SAMPLE_LEN && run.finished_at != null && run.finished_at >= run.started_at) {
      acc.durationsSecs.push(run.finished_at - run.started_at);
    }
  }
  return byRoutine;
}

/**
 * Computes every routine's reliability metrics from a fleet-wide run list, ranked worst-first:
 * an active failure streak outranks everything else (longer streak first), then lowest success
 * rate, then title for a stable tie-break. Routines with no finished run in the sample are
 * omitted — there is nothing to rank.
 */
export function computeReliability(runs: FleetRunSummary[]): RoutineReliability[] {
  const items: RoutineReliability[] = [];
  for (const [routineId, acc] of bucketFinishedRuns(runs)) {
    const sortedDurations = [...acc.durationsSecs].sort((a, b) => a - b);
    items.push({
      routineId,
      routineTitle: acc.title,
      sampleSize: acc.statuses.length,
      successes: acc.statuses.filter((s) => s === "success").length,
      streak: computeStreak(acc.statuses),
      flips: countFlips(acc.statuses),
      durationsSecs: acc.durationsSecs,
      p50Secs: percentile(sortedDurations, 0.5),
      p95Secs: percentile(sortedDurations, 0.95),
      regressing: isRegressing(acc.durationsSecs),
    });
  }

  items.sort((a, b) => {
    const streakDiff = failureStreakLen(b.streak) - failureStreakLen(a.streak);
    if (streakDiff !== 0) return streakDiff;
    const aRate = successRate(a) ?? 1;
    const bRate = successRate(b) ?? 1;
    if (aRate !== bRate) return aRate - bRate;
    return a.routineTitle.localeCompare(b.routineTitle);
  });
  return items;
}

/** Aggregates per-routine metrics into a fleet-wide summary for the page's stat tiles. */
export function fleetSummary(items: RoutineReliability[]): FleetReliability {
  const allDurations = items
    .flatMap((r) => r.durationsSecs)
    .sort((a, b) => a - b);
  return {
    sampleSize: items.reduce((sum, r) => sum + r.sampleSize, 0),
    successes: items.reduce((sum, r) => sum + r.successes, 0),
    failingCount: items.filter((r) => failureStreakLen(r.streak) > 0).length,
    flakyCount: items.filter(isFlaky).length,
    p50Secs: percentile(allDurations, 0.5),
    p95Secs: percentile(allDurations, 0.95),
    regressingCount: items.filter((r) => r.regressing).length,
  };
}

/** CSS class for a routine's streak badge (reuses the generic `run-status` pill classes). */
export function streakClass(streak: Streak): string {
  switch (streak.kind) {
    case "success":
      return "run-status success";
    case "failure":
      return "run-status failed";
    default:
      return "run-status unknown";
  }
}

/** Display label for a routine's streak badge. */
export function streakLabel(streak: Streak): string {
  switch (streak.kind) {
    case "success":
      return `${streak.count} OK`;
    case "failure":
      return `${streak.count} FAILING`;
    default:
      return "—";
  }
}

/** CSS class for a success-rate badge (reuses the generic `run-status` pill classes). */
export function rateClass(rate: number | null): string {
  if (rate === null) return "run-status unknown";
  if (rate >= 0.9) return "run-status success";
  if (rate >= 0.7) return "run-status running";
  return "run-status failed";
}

/** Display label for a success-rate badge. */
export function rateLabel(rate: number | null): string {
  return rate === null ? "—" : `${Math.round(rate * 100)}%`;
}
